File: V3_demo/api_demo/asset/Encryption_HMAC.rb
require 'digest'
require "uri"
require "net/http"
require "date"
require 'open-uri'
require 'openssl'
require 'securerandom'


$api_key='XXXXXXXXX'
$secret_key='XXXXXXXXX'
$recv_window = '5000'
$url = URI("https://api-testnet.bybit.com") # Testnet endpoint
$time_stamp = ""

def HTTP_Request(endPoint,method,payload)
    $time_stamp = DateTime.now.strftime('%Q')
    signature = genSignature(payload)
    fullUrl = $url + endPoint
    if method == "POST"
        request = Net::HTTP::Post.new(fullUrl,'Content-Type' => 'application/json')
    elsif method == "GET"
        payload="?"+payload
        fullUrl = $url + endPoint + payload
        request = Net::HTTP::Get.new(fullUrl)
    else
        puts "Check the method. It should be either GET or POST"
        exit
    end
    https = Net::HTTP.new(fullUrl.host, fullUrl.port)
    https.set_debug_output($stdout)
    https.use_ssl = true
    request["X-BAPI-API-KEY"] = $api_key
    request["X-BAPI-TIMESTAMP"] = $time_stamp
    request["X-BAPI-RECV-WINDOW"] = $recv_window
    request["X-BAPI-SIGN"] = signature
    request.body = payload
    puts payload
    response = https.request(request)
    puts response.read_body
end

def genSignature(payload)
    param_str= $time_stamp + $api_key + $recv_window + payload
    OpenSSL::HMAC.hexdigest('sha256', $secret_key, param_str)
end

#Create Internal Transfer ( SPOT to UNIFIED )
endPoint = "/asset/v3/private/transfer/inter-transfer"
method = "POST"
transferId = SecureRandom.uuid
payload='{"transferId": "' + transferId +  '","coin": "USDT","amount": "1","from_account_type": "SPOT","to_account_type": "UNIFIED"}'
HTTP_Request(endPoint,method,payload)

#Query Internal Transfer List
endPoint = "/asset/v3/private/transfer/inter-transfer/list/query"
method = "GET"
payload="coin=USDT"
HTTP_Request(endPoint,method,payload)

File: V3_demo/api_demo/asset/Encryption_HMAC.py
import requests
import time
import hashlib
import hmac
import uuid

api_key='XXXXXXXXXXXX'
secret_key='XXXXXXXXXXXX'
httpClient=requests.Session()
recv_window=str(5000)
url="https://api-testnet.bybit.com" # Testnet endpoint

def HTTP_Request(endPoint,method,payload,Info):
    global time_stamp
    time_stamp=str(int(time.time() * 10 ** 3))
    signature=genSignature(payload)
    headers = {
        'X-BAPI-API-KEY': api_key,
        'X-BAPI-SIGN': signature,
        'X-BAPI-SIGN-TYPE': '2',
        'X-BAPI-TIMESTAMP': time_stamp,
        'X-BAPI-RECV-WINDOW': recv_window,
        'Content-Type': 'application/json'
    }
    if(method=="POST"):
        response = httpClient.request(method, url+endPoint, headers=headers, data=payload)
    else:
        response = httpClient.request(method, url+endPoint+"?"+payload, headers=headers)
    print(response.text)
    print(Info + " Elapsed Time : " + str(response.elapsed))

def genSignature(payload):
    param_str= str(time_stamp) + api_key + recv_window + payload
    hash = hmac.new(bytes(secret_key, "utf-8"), param_str.encode("utf-8"),hashlib.sha256)
    signature = hash.hexdigest()
    return signature

#Create Internal Transfer ( SPOT to UNIFIED )
endpoint="/asset/v3/private/transfer/inter-transfer"
method="POST"
transferId=uuid.uuid4()
params='{"transferId": "' + str(transferId) +  '","coin": "USDT","amount": "1","from_account_type": "SPOT","to_account_type": "UNIFIED"}'
HTTP_Request(endpoint,method,params,"InternalTransfer")

#Query Internal Transfer List
endpoint="/asset/v3/private/transfer/inter-transfer/list/query"
method="GET"
params='coin=USDT'
HTTP_Request(endpoint,method,params,"InternalTransferList")
